File: readArduino.py
import serial
import time

# Initialize serial connection
def init_serial(port="/dev/ttyACM0", baud_rate=9600, timeout=1):
    try:
        ser = serial.Serial(port, baud_rate, timeout=timeout)
        print(f"Connected to {port} at {baud_rate} baud.")
        return ser
    except serial.SerialException as e:
        print(f"Failed to connect to {port}: {e}")
        return None

# Read data from Arduino
def read_from_arduino(ser):
    try:
        if ser.in_waiting > 0:
            line = ser.readline().decode('utf-8').strip()
            return float(line)
        else:
            return None
    except Exception as e:
        print(f"Error reading from serial: {e}")
        return None

# Send data to Arduino
def send_to_arduino(ser, message):
    try:
        ser.write(f"{message}\n".encode('utf-8'))
    except Exception as e:
        print(f"Error sending to Arduino: {e}")

def main():
    # Replace '/dev/ttyUSB0' with the correct port if different
    ser = init_serial(port="/dev/ttyACM0", baud_rate=9600)

    if not ser:
        return

    try:
        while True:
            sine_value = read_from_arduino(ser)
            if sine_value is not None:
                print(f"Received: {sine_value}")

                if sine_value > 100:
                    #pass
                    send_to_arduino(ser, "1")
                else:
                    #pass
                    send_to_arduino(ser, "0")
            time.sleep(0.1)

            #data = read_from_arduino(ser)
            #if data:
            #    print(f"Received: {data}")
            #time.sleep(0.1)  # Adjust for your needs
    except KeyboardInterrupt:
        print("Exiting program.")
    finally:
        if ser and ser.is_open:
            ser.close()
            print("Serial connection closed.")

if __name__ == "__main__":
    main()

File: readArduino.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <termios.h>

#define SERIAL_PORT "/dev/ttyACM0" // Replace with your port if different
#define BAUD_RATE B9600           // Baud rate for serial communication

// Function to configure the serial port
int configure_serial(const char *port) {
    int serial_fd = open(port, O_RDWR | O_NOCTTY | O_NDELAY);
    if (serial_fd == -1) {
        perror("Error opening serial port");
        return -1;
    }

    struct termios options;
    tcgetattr(serial_fd, &options);           // Get current port settings
    cfsetispeed(&options, BAUD_RATE);         // Set input baud rate
    cfsetospeed(&options, BAUD_RATE);         // Set output baud rate
    options.c_cflag |= (CLOCAL | CREAD);      // Enable receiver and set local mode
    options.c_cflag &= ~PARENB;               // No parity
    options.c_cflag &= ~CSTOPB;               // 1 stop bit
    options.c_cflag &= ~CSIZE;                // Clear current data size setting
    options.c_cflag |= CS8;                   // 8 data bits
    options.c_cflag &= ~CRTSCTS;              // Disable hardware flow control
    options.c_lflag &= ~(ICANON | ECHO | ECHOE | ISIG); // Raw input mode
    options.c_iflag &= ~(IXON | IXOFF | IXANY); // Disable software flow control
    options.c_oflag &= ~OPOST;                // Raw output mode

    tcsetattr(serial_fd, TCSANOW, &options);  // Apply settings
    return serial_fd;
}

// Function to parse comma-separated values
void parse_csv(const char *input, float *values, int max_values) {
    char *token;
    char buffer[128];
    strncpy(buffer, input, sizeof(buffer) - 1); // Copy input to buffer
    buffer[sizeof(buffer) - 1] = '\0';          // Null-terminate the buffer

    int index = 0;
    token = strtok(buffer, ",");
    while (token != NULL && index < max_values) {
        values[index] = atof(token); // Convert string to float
        token = strtok(NULL, ",");
        index++;
    }
}

int main() {
    int serial_fd = configure_serial(SERIAL_PORT);
    if (serial_fd == -1) {
        return 1; // Exit if serial port setup failed
    }

    char buffer[128];        // Buffer for incoming data
    float values[5];         // Array to store parsed values
    int bytes_read;

    printf("Listening on %s...\n", SERIAL_PORT);

    while (1) {
        // Read data from the serial port
        memset(buffer, 0, sizeof(buffer));
        bytes_read = read(serial_fd, buffer, sizeof(buffer) - 1);
        if (bytes_read > 0) {
            buffer[bytes_read] = '\0'; // Null-terminate the received string
            printf("Received: %s\n", buffer);

            // Parse the CSV data
            parse_csv(buffer, values, 5);
            printf("Parsed values: ");
            for (int i = 0; i < 5; i++) {
                printf("%.2f ", values[i]);
            }
            printf("\n");
        }

        usleep(100000); // Sleep for 100ms
    }

    close(serial_fd); // Close the serial port
    return 0;
}
